src/drivers/fluid_oboe.cpp

```diff
@@ -0,0 +1,255 @@
+  if (dev == NULL) {
+    FLUID_LOG(FLUID_ERR, "Out of memory");
+    return NULL;
+  }
+
+  FLUID_MEMSET(dev, 0, sizeof(fluid_oboe_audio_driver_t));
+  
+  dev->synth = synth;
+  dev->callback = func;
+  dev->oboe_callback = new OboeAudioStreamCallback (dev);
+
+  fluid_settings_getint(settings, "audio.period-size", &period_frames);
+  fluid_settings_getnum(settings, "synth.sample-rate", &sample_rate);
+  is_sample_format_float = fluid_settings_str_equal (settings, "audio.sample-format", "float");
+  fluid_settings_getint(settings, "audio.oboe.device-id", &device_id);
+  sharing_mode = 
+    fluid_settings_str_equal (settings, "audio.oboe.sharing-mode", "Exclusive") ? 1 : 0;
+  performance_mode =
+    fluid_settings_str_equal (settings, "audio.oboe.performance-mode", "PowerSaving") ? 1 :
+    fluid_settings_str_equal (settings, "audio.oboe.performance-mode", "LowLatency") ? 2 : 0;
+
+  builder->setDeviceId (device_id)
+	->setDirection (Direction::Output)
+	->setChannelCount (NUM_CHANNELS)
+	->setSampleRate (sample_rate)
+	->setFramesPerCallback (period_frames)
+	->setFormat (is_sample_format_float ? AudioFormat::Float : AudioFormat::I16)
+	->setSharingMode (sharing_mode == 1 ? SharingMode::Exclusive : SharingMode::Shared)
+	->setPerformanceMode (
+	  performance_mode == 1 ? PerformanceMode::PowerSaving :
+      performance_mode == 2 ? PerformanceMode::LowLatency : PerformanceMode::None)
+    ->setUsage (Usage::Media)
+    ->setContentType (ContentType::Music)
+    ->setCallback (dev->oboe_callback);
+
+  result = builder->openStream (&stream);
+  dev->stream = stream;
+  if (result != Result::OK)
+    goto error_recovery;
+
+  dev->cont = 1;
+
+  FLUID_LOG(FLUID_INFO, "Using Oboe driver");
+
+  stream->start ();
+  
+  return (fluid_audio_driver_t*) dev;
+
+ error_recovery:
+  delete_fluid_oboe_audio_driver((fluid_audio_driver_t*) dev);
+  return NULL;
+}
+
+void delete_fluid_oboe_audio_driver(fluid_audio_driver_t* p)
+{
+  fluid_oboe_audio_driver_t* dev = (fluid_oboe_audio_driver_t*) p;
+
+  if (dev == NULL) {
+    return;
+  }
+
+  dev->cont = 0;
+  
+  dev->stream->stop ();
+  
+  dev->stream->close ();
+  
+  delete dev->oboe_callback;
+  
+  FLUID_FREE(dev);
+}
+
+DataCallbackResult on_audio_ready(AudioStreamCallback *callback, AudioStream *stream, void *audioData, int32_t numFrames)
+{
+  float *callback_buffers[2];
+  fluid_oboe_audio_driver_t *dev;
+  OboeAudioStreamCallback *oboe_callback;
+  
+  oboe_callback = (OboeAudioStreamCallback*) callback;
+  dev = (fluid_oboe_audio_driver_t*) oboe_callback->user_data;
+  
+  if (!dev->cont)
+    return DataCallbackResult::Stop;
+  
+  /* FIXME: it seems that fluidsynth stores results in full float
+   * value range, while Oboe expects the results between -1 and 1.
+   * To make things worse, fluidsynth callback only accepts float*
+   * (while non-callback accepts short*) which makes it unsure if
+   * the stream is processed appropriately.
+   * So far, disable custom callback for short.
```

Comment:
That callback mode used with `new_fluid_audio_driver2()` only supports floats as buffers. [See the documentation of  fluid_audio_func_t](http://www.fluidsynth.org/api/audio_8h.html#a4c108916c658973df5242ae77fefa3d0). A custom callback mode for short isn't supported.

Reply:
Thanks, good to know.